### DISCUSSION TITLE
LiveMap JSON

### DISCUSSION THREAD

**Discussion by eben80** [dl9rdz/rdz_ttgo_sonde#226] [Developers corner]
Hi,
I know a few weeks ago there was an issue relating to the LiveMap when you are receiving multiple  sondes at once.  I couldn't imagine what it would look like since I normally never have two sondes being received at once 😄 

The way it was handled is that it was jumping between the locations because only one sonde data is available at any one time in the live.json.

Now I know that we always have the data for all received sondes because it's displayed in the data tab.

Would it be possible to have the live.json create something like below so that we can have the option of displaying multiple simultaneous or pick which sonde we would like to see tracked?

```
{
  "gps": {
    "lat": 49.0522,
    "lon": 21.282,
    "alt": 272,
    "sat": 11,
    "speed": 0.0617333,
    "dir": 0,
    "hdop": 8
  },
  "sonde": {
    "index": [
      {
        "rssi": 243,
        "vframe": 7283,
        "time": 1634907343,
        "id": "S5020910",
        "freq": 403.200,
        "type": "RS41",
        "lat": 47.868637,
        "lon": 21.345573,
        "alt": 5373,
        "dir": 73,
        "climb": -18.5,
        "speed": 25.4,
        "launchsite": "Budapest        ",
        "res": 1
      },
      {
        "rssi": 245,
        "vframe": 7688,
        "time": 1634907104,
        "id": "R5050673",
        "freq": 400.600,
        "type": "RS41",
        "lat": 49.417126,
        "lon": 22.616203,
        "alt": 30463,
        "dir": 67,
        "climb": 4.8,
        "speed": 37.8,
        "launchsite": "Poprad          ",
        "res": 1
      }
    ]
  }
}
```

Currently it's just like this:
```
{
  "sonde": {
    "rssi": 245,
    "vframe": 7688,
    "time": 1634907104,
    "id": "R5050673",
    "freq": 400.600,
    "type": "RS41",
    "lat": 49.417126,
    "lon": 22.616203,
    "alt": 30463,
    "dir": 67,
    "climb": 4.8,
    "speed": 37.8,
    "launchsite": "Poprad          ",
    "res": 1
  },
  "gps": {
    "lat": 49.0523,
    "lon": 21.282,
    "alt": 267,
    "sat": 12,
    "speed": 0.0432133,
    "dir": 0,
    "hdop": 10
  }
}
```

**Comment by dl9rdz** [comment]
I am not yet convinced this is the right way to go.

The TTGO is receiving only one sonde at a time, so there is no point in sending updates every second of all other sonde entries that certainly do not change. I see the risk of the TTGO running out of memory for having that large amount of data in network queues.

I would rather handle this on the client / javascript code side. If a new sonde is being decoded, not remove the "old" one from the map, but just keep it (maybe either with a option to manually remove it, or if it is too old). That is the way I do it on the Android app and so far nobody complained that it is not good that way ;-)

(Possibly, as extended version, the possibility to query all sonde entries on startup of the map via some request parameter)

**Comment by eben80** [comment]
Good point.
So perhaps have the additional sonde "slots" and handling of them in
Javascript on the client and update them with the current json structure.
That would also work yes.

**Comment by eben80** [comment]
@cixio would you be able to help with this handling of multiple sondes on livemap?
I have a simulator on https://ebski.co/ttgo/livemap.html
How do we use the current live.json format, store the values last values for each individual sonde, and draw multiple paths on the interface?
Would it be best to save it in the application session storage in the browser, check the sonde ID and draw concurrent paths based on that?
I suppose pruning of data older than x hours would also help.

**Reply by cixio** [reply]
I can take a look at it, but I don't have a lot of time right now.

**Reply by cixio** [reply]
Please keep the simulator running, this will help me a lot.

**Comment by cixio** [comment]
Okay, some modifications and the multi-sonde-livemap works nice, you can take a look at it in my repository: https://github.com/cixio/rdz_ttgo_sonde/commit/ec4b932c3ce7c68b133c9473e04bd822b80bee42

But now I have some problems, maybe you have an idea to solve them:
- top informations: the altitude and some more informations should be moved into the location-marker-popup to have a clear assignment. But then these infomations are not visible when you move the map. I think many users use the map to follow only one sonde, maybe this is not a good idea?
- follow sonde: should we follow the last sonde got from the ttgo? this may cause much jumps
- right button bar (location/line/prediction point): does not work - I dont know to which sonde I should move / zoom in to
- now the session storage does not clear when a new sonde is coming up - what do you think when the storage should be cleared up? After a few hours, only via button or both?

Also there is a bug with the predictions, they will be generated every 5 minutes (or earlier, depends on the height of the sonde), but when the timer pulls a new prediction only the prediction for the last sonde got from the ttgo will be updated. Need some time to fix it, because I dont know how to manage the different update-times.

**Reply by eben80** [reply]
Hey I had no idea how to draw the two lines so that's progress already. I've updated the demo page with your code so long.
https://ebski.co/ttgo/livemap.html
Maybe if we create a listing of received sondes with checkboxes to toggle their display, radio button for more info and to purge them somehow bottom left?
The sonde selected for detail display will then show details in the top center... Just some idea.

Perhaps first behaviour is to zoom to both paths bound box so it doesn't jump.

**Reply by cixio** [reply]
Fixed right button bar - shows the details of the latest sonde.

https://github.com/cixio/rdz_ttgo_sonde/commit/864baa5c882799639ad64a5ad13cc4ae10715cc8

**Reply by eben80** [reply]
I've changed the code on the simulator to allow you to cycle between light, dark and satellite maps, irrespective of the system or user selected theme.

I found that I prefer the light mode map, even if I have the dark mode on my device so this gives the option to choose that.
Check it out on the simulator: [https://ebski.co/ttgo/livemap.html](https://ebski.co/ttgo/livemap.html)
What do you think?
